File: src/leetcode/python/medium/FlattenBinaryTreeToLinkedList_114.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class FlattenBinaryTreeToLinkedList_114(object):
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: void Do not return anything, modify root in-place instead.
        """
        if not root:
            return
        if not root.left and not root.right:
            return

        self.flatten(root.right)
        self.flatten(root.left)
        if root.left and root.right:
            pre = root.left
            cur = pre.right
            while cur:
                pre = cur
                cur = pre.right
            pre.right = root.right
            root.right = root.left
            root.left = None
        if root.left and not root.right:
            root.right = root.left
            root.left = None

File: src/leetcode/java/FlattenBinaryTreeToLinkedList_114.java
package leetcode.java;

public class FlattenBinaryTreeToLinkedList_114 {
  public void flatten(TreeNode root) {
    if (root == null) return;
    if (root.left == null && root.right == null) return;

    flatten(root.right);
    flatten(root.left);
    if (root.left != null && root.right != null) {
      TreeNode pre = root.left;
      TreeNode cur = pre.right;
      while (cur != null) {
        pre = cur;
        cur = pre.right;
      }
      pre.right = root.right;
      root.right = root.left;
      root.left = null;
    }
    if (root.left != null && root.right == null) {
      root.right = root.left;
      root.left = null;
    }
  }

  class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
  }
}
